src/domain_bridge/domain_bridge.cpp

```diff
@@ -0,0 +1,218 @@
+// Copyright 2021, Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "domain_bridge/domain_bridge.hpp"
+
+#include <map>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+
+#include "domain_bridge/exceptions.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/multi_threaded_executor.hpp"
+#include "rcutils/logging_macros.h"
+#include "rosbag2_cpp/typesupport_helpers.hpp"
+
+#include "generic_publisher.hpp"
+#include "generic_subscription.hpp"
+
+namespace domain_bridge
+{
+
+/// Implementation of \ref DomainBridge.
+class DomainBridgeImpl
+{
+public:
+  using SubscriptionMap =
+    std::map<std::pair<size_t, std::string>, std::shared_ptr<GenericSubscription>>;
+  using PublisherMap = std::map<std::pair<size_t, std::string>, std::shared_ptr<GenericPublisher>>;
+  using NodeMap = std::map<size_t, std::shared_ptr<rclcpp::Node>>;
+
+  DomainBridgeImpl()
+  {
+    this->executor_context_ = create_context_with_domain_id(1);
+    rclcpp::ExecutorOptions executor_options;
+    executor_options.context = this->executor_context_;
+    this->executor_ = std::make_unique<rclcpp::executors::MultiThreadedExecutor>(executor_options);
+  }
+
+  ~DomainBridgeImpl()
+  {
+    const std::string reason("Domain bridge shutdown");
+    this->executor_context_->shutdown(reason);
+  }
+
+  rclcpp::Context::SharedPtr create_context_with_domain_id(size_t domain_id)
+  {
+    auto context = std::make_shared<rclcpp::Context>();
+    rclcpp::InitOptions options;
+    options.auto_initialize_logging(false).set_domain_id(domain_id);
+    context->init(0, nullptr, options);
+    return context;
+  }
+
+  rclcpp::NodeOptions create_node_options(rclcpp::Context::SharedPtr context)
+  {
+    rclcpp::NodeOptions options;
+    return options.context(context)
+           .use_global_arguments(false)
+           .start_parameter_services(false)
+           .start_parameter_event_publisher(false);
+  }
+
+  rclcpp::Node::SharedPtr get_node_for_domain(size_t domain_id)
+  {
+    auto domain_id_node_pair = node_map_.find(domain_id);
+
+    // If we don't already have a node for the domain, create one
+    if (node_map_.end() == domain_id_node_pair) {
+      auto context = create_context_with_domain_id(domain_id);
+      auto node_options = create_node_options(context);
+      std::ostringstream oss;
+      oss << "domain_bridge_" << std::to_string(domain_id);
```

Comment:
There could be potentially more than one domain bridge.
I would use something like: `domain_bridge_{domain_bridge_id}_{domain_id}`

Where the "domain_bridge_id" is a random number representing the bridge (or anything that works a pseudo-unique id for the "bridge").

Reply:
I've made this ID configurable via a new options structure: b39167a

This way users can get a deterministic output for the node names.